rsp_jupyter_extensions/handlers/tutorials.py

```diff
@@ -0,0 +1,413 @@
+            h.entries = {}
+        h.entries[nm] = h_entry
+    return h
+
+
+# _find_repo and _get_tag might belong in lsst.rsp.
+
+
+def _find_repo() -> str | None:
+    # Eventually we may not even want to set AUTO_REPO_SPECS...
+    auto_repos = os.getenv("AUTO_REPO_SPECS")
+    if not auto_repos:
+        # This instance doesn't want them
+        return None
+    repos = auto_repos.split(",")
+    for repo in repos:
+        if repo.find("tutorial-notebooks") > -1:
+            break
+    atsign = repo.find("@")
+    if atsign == -1:
+        repo += "@main"
+    return repo
+
+
+def _get_tag() -> str:
+    image = os.getenv("JUPYTER_IMAGE_SPEC")
+    if not image:
+        raise UserEnvironmentError(
+            "Environment variable 'JUPYTER_IMAGE_SPEC' is not set"
+        )
+    colon = image.find(":")
+    atsign = image.find("@")
+    if colon < 0 or atsign <= colon:
+        raise TagError("Could not extract tag from image spec")
+    tag = image[colon + 1 : atsign]
+    if not tag:
+        raise TagError("Could not determine image tag")
+    return tag
+
+
+# Generic clone method
+
+
+def _clone_repo(repo_url: str, branch: str, dirname: str) -> None:
+    runner = cmd.Command()
+    proc = runner.run(
+        "git",
+        "clone",
+        "--depth",
+        "1",
+        repo_url,
+        "-b",
+        branch,
+        dirname,
+        timeout=30,
+    )
+    if proc.returncode != 0:
+        raise RuntimeError(f"git clone {repo_url}@{branch} failed")
+
+
+def _get_homedir() -> Path:
+    homedir = os.getenv("HOME")
+    if not homedir:
+        raise UserEnvironmentError("home directory is not set")
+    return Path(homedir)
+
+
+# RSP-specific tutorial locations
+
+
+def _check_tutorials_hierarchy_stash() -> Hierarchy | None:
+    # This is a little subtle.  We get a new instance of the handler with
+    # every access of its endpoints, and however Jupyter Server manages that
+    # it really needs to be a new one--making the handler a singleton does
+    # not work (it doesn't respond to its endpoints).
+    #
+    # So what we do is to check the presence of a serialized hierarchy
+    # in a known location inside the user's homedir.  If it exists and it
+    # is sufficiently new (let's start with 8 hours or less) then we
+    # deserialize it and return that.
+    #
+    # If it does not exist or is potentially stale...we return None from
+    # here, which then allows the clone to proceed, and we rebuild the stash
+    # from those results.
+    max_age = datetime.timedelta(hours=8)
+    homedir = _get_homedir()
+    stash = homedir / ".cache" / "tutorials.json"
+    if not stash.is_file():
+        return None
+    mod = datetime.datetime.fromtimestamp(
+        stash.stat().st_mtime, tz=datetime.UTC
+    )
+    now = datetime.datetime.now(tz=datetime.UTC)
+    age = now - mod
+    if age > max_age:
+        return None
+    return Hierarchy.from_primitive(json.loads(stash.read_text()))
```

Comment:
To be extra safe would it make sense to catch errors loading the JSON file (if it is corrupted or modified for some reason to a non valid json file) and then return `None`, so the hierarchy is repopulated?

```suggestion
    try:
        content = stash.read_text()
        json_data = json.loads(content)
        return Hierarchy.from_primitive(json_data)
    except (OSError, json.JSONDecodeError, HierarchyError):
        return None
```